### ISSUE TITLE
Warn admins when they've added an API backend that is un-routable due to prefix matching order

### ISSUE THREAD

**Issue by GUI** [18F/api.data.gov#186]
This has come up a couple times this week: Let's say as an admin, you've added two different API Backend configurations. The first API backend you add has a matching URL prefix of `/nrel/`. Since this is a prefix matcher, that means any URLs starting with `/nrel/*` will get routed to this API backend. Then let's say you add a second API backend with a URL prefix of `/nrel/staging/`. The problem is that this second API backend will be unroutable by default. Since the first API backend's prefix is broader than the second, a request to `/nrel/staging/foo` will always be matched by the first `/nrel/` prefix and therefore never make it to the more specific staging backend that got added second.

This matching order issue is a very solvable problem in the current system, but if you're not aware of this matching order business, it can be quite unintuitive why your newly added API backend doesn't seem to be routing like you expected. What I'd like to do is add a warning if you add a new API backend that won't be matched by default. We could then recommend the couple of ways you could fix this. I think determining unroutable APIs should be relatively straight-forward: I think we just need to walk all of the prefixes in matching order and then warn if we see one that has a more specific prefix than one we've already seen.

**Comment by GUI** [comment]
For more general reference, the couple ways to fix this are:

- Re-sort the API backends, so the `/nrel/staging/` one would be matched before the broader `/nrel/` (there's a "Reorder" button on the listing page that allows you to drag things around to reorder them).
- If possible, flip around your URL structure to eliminate possible conflicts. So instead of using `/nrel/staging/`, use `/staging/nrel/`. That way, you have no possible conflicts. This isn't always possible depending on your URL requirements, but I do tend to favor this approach a bit for broader setups like staging, since it eliminates the possibility of conflicts.

But I'd like to tackle this in conjunction with #187. Because if we start making it more obvious to admins that they can reorder APIs, then I'd like to get a better handle on the reordering algorithm, so we're not causing lots of API backend changes.

**Comment by GUI** [comment]
We've rolled out changes that will hopefully eliminate the need for this by making the path match ordering more intuitive. It's essentially what was outlined in https://github.com/NREL/api-umbrella/issues/342#issuecomment-287622098, where we always match more specific URL prefixes (ie, longer path prefixes) first. This way if you add `/foo` and `/foo/bar`, `/foo/bar` will always match before `/foo`, regardless of what order you added these in. This eliminates the need to manually sort API Backends and Matching URL Prefixes entirely. In cases where URL prefixes conflict across different API backends, the first API backend added will take precedence. Sub-URL Request Settings and Advanced Requests Rewriting are still manually sortable, since these rely on regex matches, where we don't have an automatic way to determine what should take precedence.

When migrating our database to this new strategy, we identified several path prefixes that would have led to different routing with the new rules in place. This was due to situations where API backend were published with un-routable path orders (eg, if it had published both `/foo` and `/foo/bar` in that order, `/foo/bar` would never be matched). In order to maintain compatibility with what was previously live, the database migration removed these un-matchable (but more specific) URL prefixes from the published API config. But the configuration remains in the database, so in these cases agencies will now see pending changes to be published where the diff will reveal the additional URL matches that would be added that would cause different routing results.
